### PR TITLE
Workflows holdingpen changes

### PR DESCRIPTION
**Pull request by jalavik**
Hello:

Here are the latest changes to the Holding Pen that @dinossimpson have been working on. I merged it with the new structure and I hope it works (no way of testing the UI due to my current problems, see #49).

I discovered that in these changes we do need autodiscovery of widgets in the holding pen. So for now I kinda reverted to the old method as I am not sure which method is going to be the preferred way in next.

Cheers,
Jan

### CODE REVIEW

File: modules/bibworkflow/lib/bibworkflow_utils.py [hunk 1]
```diff
@@ -210,3 +236,121 @@ def set_up_redis():
...
     )
     return redis_server
+
+
+def empty_redis():
+    redis_server = set_up_redis()
+    redis_server.flushall()
+
+
+def pretty_date(time=False):
```
Comment (jirikuncar) [inline modules/bibworkflow/lib/bibworkflow_utils.py]: There is already: `pretty_date` in `inveino.utils.date`.

File: modules/bibworkflow/etc/templates/bibworkflow_hp_maintable.html [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    
+{%- js 'js/jquery.dataTables.min.js' -%}
+{%- js 'js/DT_bootstrap.js' -%}
+{%- js 'js/bibworkflow_hp_maintable.js' -%}
+{%- js 'js/ColVis.js' -%}
+{%- js 'js/ColVis.min.js' -%}
+
+
+    <style type="text/css" title="currentStyle">
+        @import "../../css/DT_bootstrap.css";
```
Comment (jirikuncar) [inline modules/bibworkflow/etc/templates/bibworkflow_hp_maintable.html]: Use `url_for('workflows.static', filename='...')`

**Comment by jalavik** [comment]
I have updated the branch now. However, when testing the web interface things are not working so well.

I have problems adding the module JS in the Holding Pen interfaces (FIXED). Secondly I also have trouble making the workflows admin interface work. I get a 404 on loading /admin/workflows.

Probably these problems are easy to fix and just due to my lack of experience with jinja/flask.

File: invenio/modules/workflows/static/js/workflows/hp_datapreview.js [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+// along with Invenio; if not, write to the Free Software Foundation, Inc.,
+// 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.
+
+$(document).ready(function(){    
+    var bwoid;
+    var datapreview = "hd";
+
+    window.data_preview = function(format){
+        jQuery.ajax({
+            url: "/admin/holdingpen/entry_data_preview?oid="+bwoid+"&recformat="+format,
```
Comment (jirikuncar) [inline invenio/modules/workflows/static/js/workflows/hp_datapreview.js]: url should be configurable

File: invenio/modules/workflows/templates/workflows/hp_details.html [hunk 4]
```diff
@@ -1,12 +1,18 @@
...
 
-    <h1>Record Details</h1>
+    <div class="row-fluid">
+        <div class="span3 pull-left"><a href="{{url_for('holdingpen.maintable')}}"><i class="icon-hand-left"></i> Back to Main Table</a></div>
+        <!-- check with Jan | debate if we should have headers in pages in HP -->
+        <!-- <div class="span3"><h1>Record Details</h1></div> -->
+    </div>
+
+    <!-- FIXME: temporary external loading of prettify -->
     <script src="https://google-code-prettify.googlecode.com/svn/loader/run_prettify.js?lang=css&amp;skin=desert"></script>
```
Comment (jirikuncar) [inline invenio/modules/workflows/templates/workflows/hp_details.html]: Why are you using this external code?

File: invenio/modules/workflows/templates/workflows/hp_index.html [hunk 5]
```diff
@@ -1,50 +1,23 @@
 {% extends "workflows/hp_layout.html" %}
 {% block body2 %}
```
Comment (jirikuncar) [inline invenio/modules/workflows/templates/workflows/hp_index.html]: `body2` is not the best name for block, can you find different one?

File: invenio/modules/workflows/views/holdingpen.py [hunk 6]
```diff
@@ -17,130 +17,218 @@
...
 from invenio.base.decorators import templated, wash_arguments
+from invenio.modules.formatter.engine import format_record
 from invenio.base.i18n import _
 from invenio.ext.breadcrumb import default_breadcrumb_root, register_breadcrumb
 from invenio.ext.menu import register_menu
-from invenio.bibworkflow_utils import get_workflow_definition
-from invenio.bibworkflow_api import continue_oid_delayed
-from invenio.bibworkflow_config import CFG_OBJECT_VERSION
+from invenio.utils.date import pretty_date
+from invenio.bibworkflow_utils import (get_workflow_definition,
```
Comment (jirikuncar) [inline invenio/modules/workflows/views/holdingpen.py:31]: Can you move `bibworkflow_{api,config,utils}` to `invenio/modules/workflows/{api,config,utils}`?

File: invenio/modules/workflows/views/holdingpen.py [hunk 7]
```diff
@@ -17,130 +17,218 @@
...
 @register_menu(blueprint, 'main.admin.holdingpen', _('Holdingpen'))
+@register_breadcrumb(blueprint, '.', _('Holdingpen'))
 @templated('workflows/hp_index.html')
 def index():
     """
-    Displays main interface of BibHoldingpen.
+    Displays main interface of Holdingpen.
+    Acts as a hub for catalogers (may be removed)
+    """
+    from invenio.bibworkflow_containers import bwolist
```
Comment (jirikuncar) [inline invenio/modules/workflows/views/holdingpen.py:54]: Can you move `bibworkflow_containers` to `invenio/modules/workflows/containers`?

**Comment by jalavik** [comment]
Yep, sure. I'll take a look at these. I wanted to move them, but right after, danger of doing too much at the same time.

First I wanted to get the holding pen web interfaces to work overall. This I am getting close to - found out where the error of JS inclusion was.

File: invenio/modules/workflows/views/holdingpen.py [hunk 8]
```diff
@@ -17,130 +17,218 @@
...
-    iDisplayLength = request.args.get('iDisplayLength')
     # sSearch will be used for searching later
-    # sSearch = request.args.get('sSearch')
-    # iSortCol_0 = request.args.get('iSortCol_0')
-    # sSortDir_0 = request.args.get('sSortDir_0')
+    sSearch = request.args.get('sSearch')
+    for widget in widgets:
+        if widget == sSearch:
+            sSearch = widget
+    iSortCol_0 = request.args.get('iSortCol_0')
```
Comment (jirikuncar) [inline invenio/modules/workflows/views/holdingpen.py]: is it possible to use lower cased arguments?

File: invenio/modules/workflows/views/holdingpen.py [hunk 9]
```diff
@@ -149,42 +237,23 @@ def details(bwobject_id):
     Displays info about the hpcontainer, and presents the data
     of all available versions of the object. (Initial, Error, Final)
     """
-    # search for parents
-    bwobject = BibWorkflowObject.query.filter(BibWorkflowObject.id ==
-                                              bwobject_id).first()
+    current_app.config['breadcrumbs_map'][request.endpoint] = \
```
Comment (jirikuncar) [inline invenio/modules/workflows/views/holdingpen.py]: This is not working anymore. Look to `invenio.modules.search.views.search` how to append breadcrumbs.

**Comment by jalavik** [comment]
Oh, and I also found a bug in utils.date that I will push a fix for shortly.

File: invenio/modules/workflows/widgets/approval_widget.py [hunk 10]
```diff
@@ -0,0 +1,82 @@
+# -*- coding: utf-8 -*-
```
Comment (jirikuncar) [inline invenio/modules/workflows/widgets/approval_widget.py:1]: Be careful about naming it just widgets. Are you sure that they are general reusable WTForms?

**Comment by jirikuncar** [comment]
> I have problems adding the module JS in the Holding Pen interfaces. Secondly I also have trouble making the workflows admin interface work. I get a 404 on loading /admin/workflows.

Look that the view is loaded in `invenio.modules.workflows.views` and it's importable.

File: invenio/modules/workflows/templates/workflows/hp_details.html (hunk 4 above)
Comment (jalavik) [inline invenio/modules/workflows/templates/workflows/hp_details.html]: So there was some troubles making it work locally for some unknown reason, but we'll look into it again, and see if we can get it to work properly.

File: invenio/modules/workflows/views/holdingpen.py (hunk 8 above)
Comment (jalavik) [inline invenio/modules/workflows/views/holdingpen.py]: Done

File: invenio/modules/workflows/views/holdingpen.py (hunk 9 above)
Comment (jalavik) [inline invenio/modules/workflows/views/holdingpen.py]: Ok, temp fix done. We will figure out the dynamic parts later

File: invenio/modules/workflows/templates/workflows/hp_details.html (hunk 4 above)
Comment (jalavik) [inline invenio/modules/workflows/templates/workflows/hp_details.html]: Added and working as library

File: invenio/modules/workflows/views/holdingpen.py (hunk 8 above)
Comment (jalavik) [inline invenio/modules/workflows/views/holdingpen.py]: The CamelCase was there as this is how the DataTables API reference its variables. However, we did rename the Python variables properly now.

**Comment by jirikuncar** [comment]
Thanks for `formatter` and `date` fixes. Please create another pull request when you move `bibworkflow` to `workflows`.

**Pull request merged**